Automattic/pocket-casts-ios
# End of Year: preload stories

| 📘 Project: #376 |
|:---:|

To provide a better experience for the users, the app preloads the next 2 stories. The assets that are preloaded are:

1. Image covers
2. Podcast colors

Those are requested if they do not exist, and they are requests only once (images by using Kingfisher and colors by using the current `ColorManager`).

This also changes how the `StoriesView` is rendered to avoid Stories being re-created hundreds of time.

## To test

### Playing stories without skipping to next/previous

1. Enable the `endOfYear` flag in `FeatureFlag.swift`
2. Do a fresh install of the app
3. Login to an account with a few items on Listening History
3. Use a tool like Network Link Conditioner to change the speed of your connection (to 3G, for example)
3. When the prompt appears, tap "View My 2022"
3. ✅ Leave the stories to play and check that most stuff should be ready when the stories appear

### Skipping

1. Enable the `endOfYear` flag in `FeatureFlag.swift`
2. Do a fresh install of the app
3. Login to an account with a few items on Listening History
3. Use a tool like Network Link Conditioner to change the speed of your connection (to 3G, for example)
3. When the prompt appears, tap "View My 2022"
3. ✅ Quickly goes to the next story and you should be able to see that the background is opaque but once colors are downloaded it changes

### Number of Stories View

A bit of context: `StoriesModel` had the `progress` property as a published one. Each time that changed, `StoriesView` was invalidated and re-rendered again. In turn, this caused the current story to be invalidated and re-rendered again.

In terms of allocation, this wasn't a big problem because the views are mainly `structs`, but the CPU usage was going up due to the view being invalidated every 0.01s.

I moved the progress to another object that is used only by the progress indicators, now the `StoriesView` is only invalidated when the current story changes.

To test:

1. Enable the `endOfYear` flag in `FeatureFlag.swift`
2. Run the app and make sure you're logged in to an account with a few items on Listening History
3. Run Instruments > Allocations and start recording the current Pocket Casts instance
4. Add "Story" to the "Detail Filter"
3. Check that only 1 to 2 total Story views area created

## Checklist
